pkg/auth/service.go

```diff
@@ -474,299 +473,275 @@ func (s *DBAuthService) ListEffectivePolicies(ctx context.Context, username stri
-		}
-		p := &model.Paginator{}
-		if len(groups) == params.Amount+1 {
-			// we have more pages
-			groups = groups[0:params.Amount]
-			p.Amount = params.Amount
-			p.NextPageToken = groups[len(dbGroups)-1].DisplayName
-			return &res{groups, p}, nil
-		}
-		p.Amount = len(groups)
-		return &res{groups, p}, nil
-	})
 
+	gu := model.KvUserToGroup(username, groupDisplayName)
+	err := s.store.DeleteMsg(ctx, model.PartitionKey, gu)
 	if err != nil {
-		return nil, nil, err
+		return fmt.Errorf("remove user from group: (key %s): %w", gu, err)
 	}
-	return result.(*res).groups, result.(*res).paginator, nil
+	return err
 }
 
-func (s *DBAuthService) ListGroupUsers(ctx context.Context, groupDisplayName string, params *model.PaginationParams) ([]*model.User, *model.Paginator, error) {
-	type res struct {
-		users     []*model.User
-		paginator *model.Paginator
+func (s *KVAuthService) ListUserGroups(ctx context.Context, username string, params *model.PaginationParams) ([]*model.Group, *model.Paginator, error) {
+	if _, err := s.GetUser(ctx, username); err != nil {
+		return nil, nil, fmt.Errorf("%s: %w", username, err)
 	}
-	result, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		if _, err := getGroup(tx, groupDisplayName); err != nil {
-			return nil, err
-		}
-		dbUsers := make([]*model.DBUser, 0)
-		err := tx.Select(&dbUsers, `
-			SELECT auth_users.* FROM auth_users
-				INNER JOIN auth_user_groups ON (auth_users.id = auth_user_groups.user_id)
-				INNER JOIN auth_groups ON (auth_user_groups.group_id = auth_groups.id)
-			WHERE
-				auth_groups.display_name = $1
-				AND auth_users.display_name > $2
-				AND auth_users.display_name like $3
-			ORDER BY auth_groups.display_name
-			LIMIT $4`,
-			groupDisplayName, params.After, fmt.Sprint(params.Prefix, "%"), params.Amount+1)
+
+	hasMoreGroups := true
+	nextPage := ""
+	count := 0
+	userGroups := make([]*model.Group, 0)
+	for hasMoreGroups && count < params.Amount {
+		groups, paginator, err := s.ListGroups(ctx, params)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
+		}
+		for _, group := range groups {
+			path := model.KvUserToGroup(username, group.DisplayName)
+			m := model.UserData{}
+			_, err := s.store.GetMsg(ctx, model.PartitionKey, path, &m)
+			if err != nil && !errors.Is(err, kv.ErrNotFound) {
+				return nil, nil, err
+			}
+			if !errors.Is(err, kv.ErrNotFound) {
+				userGroups = append(userGroups, group)
+				count += 1
+			}
 		}
-		p := &model.Paginator{}
-		users := make([]*model.User, len(dbUsers))
-		for i := range dbUsers {
-			users[i] = model.ConvertUser(dbUsers[i])
+		if paginator.NextPageToken == "" {
+			hasMoreGroups = false
 		}
-		if len(users) == params.Amount+1 {
-			// we have more pages
-			users = users[0:params.Amount]
-			p.Amount = params.Amount
-			p.NextPageToken = users[len(users)-1].Username
-			return &res{users, p}, nil
+		params = &model.PaginationParams{
+			After: paginator.NextPageToken,
```

Comment:
What about the other properties of `params`? Does it make sense to fetch for the default amount when much less was requested?

Reply:
Irrelevant, not the latest version.